test/unit/matrix/test_matrix.cpp

```diff
@@ -528,64 +686,112 @@ TYPED_TEST(MatrixTest, Dependencies) {
+      }();
+
+      auto rosenders4a = getReadSendersUsingGlobalIndex(mat);
+      EXPECT_TRUE(checkSendersStep(0, rosenders4a));
+
+      CHECK_MATRIX_SENDERS(true, senders1, senders0);
+      EXPECT_TRUE(checkSendersStep(0, rosenders2b));
+      CHECK_MATRIX_SENDERS(true, rosenders2b, senders1);
+      EXPECT_TRUE(checkSendersStep(rosenders2a.size(), rosenders2a));
+
+      CHECK_MATRIX_SENDERS(false, senders3, rosenders2b);
+      CHECK_MATRIX_SENDERS(true, senders3, rosenders2a);
+
+      CHECK_MATRIX_SENDERS(true, rosenders4a, senders3);
+
+      auto rosenders4b = getReadSendersUsingLocalIndex(mat);
+      EXPECT_TRUE(checkSendersStep(rosenders4b.size(), rosenders4b));
+
+      auto senders5 = getReadWriteSendersUsingGlobalIndex(mat);
+      EXPECT_TRUE(checkSendersStep(0, senders5));
+
+      CHECK_MATRIX_SENDERS(false, senders5, rosenders4a);
+      CHECK_MATRIX_SENDERS(true, senders5, rosenders4b);
     }
   }
 }
 
-TYPED_TEST(MatrixTest, DependenciesReferenceMix) {
+TYPED_TEST(MatrixTest, DependenciesSubSubPipeline) {
   using Type = TypeParam;
 
   for (const auto& comm_grid : this->commGrids()) {
     for (const auto& test : sizes_tests) {
       // Dependencies graph:
-      // rw0 - rw1 - ro2a - rw3 - ro4a - rw5
-      //           \ ro2b /    \ ro4b /
+      // rw0 - rw1 - ro2a ------- rw3 - ro4a ------- rw5
+      //           \ ------ ro2b /     \ ----- ro4b /
+      //
+      //             +---------------------+
+      //                  sub pipeline
+      //
+      //                    +-------+
+      //                sub-sub pipeline
 
       GlobalElementSize size = globalTestSize(test.size, comm_grid.size());
       Matrix<Type, Device::CPU> mat(size, test.block_size, comm_grid);
 
-      auto senders0 = getReadWriteSendersUsingGlobalIndex(mat);
+      auto senders0 = getReadWriteSendersUsingLocalIndex(mat);
       EXPECT_TRUE(checkSendersStep(senders0.size(), senders0));
 
-      auto senders1 = getReadWriteSendersUsingLocalIndex(mat);
+      auto senders1 = getReadWriteSendersUsingGlobalIndex(mat);
       EXPECT_TRUE(checkSendersStep(0, senders1));
 
-      auto rosenders2a = getReadSendersUsingGlobalIndex(mat);
-      EXPECT_TRUE(checkSendersStep(0, rosenders2a));
+      auto [rosenders2a, rosenders2b, senders3, rosenders4a] = [&]() {
+        auto mat_sub = mat.subPipeline();
 
-      decltype(rosenders2a) rosenders2b;
-      {
-        Matrix<const Type, Device::CPU>& const_mat = mat;
-        rosenders2b = getReadSendersUsingLocalIndex(const_mat);
-        EXPECT_TRUE(checkSendersStep(0, rosenders2b));
-      }
+        auto rosenders2a = getReadSendersUsingLocalIndex(mat_sub);
+        EXPECT_TRUE(checkSendersStep(0, rosenders2a));
 
-      auto senders3 = getReadWriteSendersUsingGlobalIndex(mat);
-      EXPECT_TRUE(checkSendersStep(0, senders3));
+        auto [rosenders2b, senders3] = [&]() {
+          auto rosenders2b = getReadSendersUsingGlobalIndex(mat_sub);
```

Comment:
You're right, good catch!

Reply:
See https://github.com/eth-cscs/DLA-Future/pull/898/commits/8511a70c1ebde847bb251602ca8e195aa1ff9770. I had to update the test a bit because of the missing read-after-read optimization.